### PR TITLE
Add a flag to detect access to grad before calling `cleargrad`

### PR DESCRIPTION
**Pull request by niboshi**
Currently `Parameter.grad` is initialized with NaN if `cleargrad` is not called. This is to notify users that they mistakenly forgot to call `cleargrad`.

This PR is an attempt to detect that by a flag.

### CODE REVIEW

File: chainer/variable.py [hunk 1]
```diff
@@ -1932,10 +1948,12 @@ def as_array(obj):
     return obj
 
 
-def _recover_parameter(data, name, grad, initializer, update_rule, device):
+def _recover_parameter(
+        data, name, grad, grad_valid, initializer, update_rule, device):
```
Comment (niboshi) [inline chainer/variable.py]: Does it cause snapshot incompatibility?

File: chainer/variable.py [hunk 2]
```diff
@@ -948,6 +963,12 @@ def grad(self):
         and error.
 
         """
+        if not self._grad_valid:
+            raise RuntimeError(
+                'Cannot retrieve Variable.grad because it\'s not valid. '
```
Comment (Crissman) [inline chainer/variable.py]:
```suggestion
                'Cannot retrieve Variable.grad because it must be set '
```

File: chainer/variable.py [hunk 3]
```diff
@@ -948,6 +963,12 @@ def grad(self):
...
+                'Either it must be set beforehand or Variable.cleargrad() '
```
Comment (Crissman) [inline chainer/variable.py]:
```suggestion
                'beforehand or Variable.cleargrad() must be called.')
```

File: chainer/variable.py [hunk 4]
```diff
@@ -948,6 +963,12 @@ def grad(self):
...
+                'must be called.')
```
Comment (Crissman) [inline chainer/variable.py]:
``````

**Comment by Crissman** [comment]
Please look at comments and resolve conflicts. 🙇

File: chainer/variable.py (hunk 2 above)
Comment (niboshi) [reply chainer/variable.py]: How about "... because neither it's set beforehand nor Variable.cleargrad() is called" ?

Comment (Crissman) [reply chainer/variable.py]: `Neither... nor...` is difficult to understand English. I think my first revision is more clear.

Comment (niboshi) [reply chainer/variable.py]: Hmm.
I feel your version sounds to suggest the cause is "it must be ..." and does not state whether "it actually is".

Comment (niboshi) [reply chainer/variable.py]: How about removing "because it's not valid" from the original version, leaving two sentences?

Comment (Crissman) [reply chainer/variable.py]: So,
```Cannot retrieve Variable.grad. Either it must be set beforehand or Variable.cleargrad() must be called.```

I'm ok with that.

Comment (Crissman) [reply chainer/variable.py]: Actually, for more clarity, should be,

```Cannot retrieve Variable.grad. Either it must be set manually or Variable.cleargrad() must be called beforehand.```

Comment (niboshi) [reply chainer/variable.py]: Thank you for correction. Fixed as such.

**Comment by Crissman** [comment]
Jenkins, test this please.

**Comment by Crissman** [comment]
Jenkins failure looks unrelated.

**Pull request merged**